library/gcm.c

```diff
@@ -442,6 +464,11 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
         return( MBEDTLS_ERR_GCM_BAD_INPUT );
     }
 
+    if ( ( ctx->len == 0 ) && ( ctx->add_len % 16 ) )
+    {
```

Comment:
- [ ]  add a test function with non-empty AD and empty encrypted data, checking both encryption and decryption, and calling mbedtls_gcm_update(0) 0 or 1 time based on the value of a parameter.

Reply:
Hm. I'm not sure I did what you expected. I added a test in the test data, to cover this scenario. Is it ok?

Reply:
I meant to test the scenario where the functions are called in the sequence above. There is currently no test scenario where `mbedtls_gcm_update_ad` is called at least once, but `mbedtls_gcm_update` is not called at all. When we test, even if the plaintext/ciphertext is empty, we call `mbedtls_gcm_update` with a length of 0. It might well make a difference whether `mbedtls_gcm_update` is called at all or not: a plausible bug would be that the last partial block of AD is processed correctly only if `mbedtls_gcm_update` is called at least once, or only if `mbedtls_gcm_update` is called at most once, or only if `mbedtls_gcm_update` is not called at all.

So we need a new test function, which is specialized to empty plaintext/ciphertext, where the AD is a test parameter and where the number of times to call `mbedtls_gcm_update` is a parameter or a loop index. This must be called for at least AD lengths 0, 16 and some number not multiple of 16; for both encryption and decryption; with `mbedtls_gcm_update` called 0, 1 and 2 times.

Reply:
And similarly we should have tests with empty AD where `mbedtls_gcm_update_ad` is not called at all. Given the current structure of the code, a bug in this case is not likely, but we should test it in case the structure of the code changes.